Library/Homebrew/utils/ruby.sh

```diff
@@ -0,0 +1,46 @@
+setup-ruby-path() {
+  local HOMEBREW_VENDOR_RUBY_OPT;
+  local HOMEBREW_VENDOR_RUBY_PATH;
+
+  if [[ -z "$HOMEBREW_DEVELOPER" ]]
+  then
+    unset HOMEBREW_RUBY_PATH
+  fi
+
+  if [[ -z "$HOMEBREW_RUBY_PATH" && "$HOMEBREW_COMMAND" != "install-vendor" ]]
+  then
+    HOMEBREW_VENDOR_RUBY_OPT="$HOMEBREW_LIBRARY/Homebrew/vendor/ruby/opt"
+    HOMEBREW_VENDOR_RUBY_PATH="$HOMEBREW_VENDOR_RUBY_OPT/bin/ruby"
+    if [[ -x "$HOMEBREW_VENDOR_RUBY_PATH" ]]
+    then
+      HOMEBREW_RUBY_PATH="$HOMEBREW_VENDOR_RUBY_PATH"
+
+      if [[ $(readlink "$HOMEBREW_VENDOR_RUBY_OPT") != "$(<"$HOMEBREW_LIBRARY/Homebrew/vendor/ruby-version")" ]]
+      then
+        if ! brew install-vendor ruby --quiet
+        then
+          onoe "Failed to upgrade vendor Ruby."
+        fi
+      fi
+    else
+      if [[ -n "$HOMEBREW_OSX" ]]
+      then
+        HOMEBREW_RUBY_PATH="/System/Library/Frameworks/Ruby.framework/Versions/Current/usr/bin/ruby"
+      else
+        HOMEBREW_RUBY_PATH="$(which ruby)"
+      fi
+
+      if [[ -z "$HOMEBREW_RUBY_PATH" || "$("$HOMEBREW_RUBY_PATH" -e "puts RUBY_VERSION.split('.').first")" != "2" ]]
```

Comment:
My very simplistic experiment yielded about 10 ms for `ruby --version` and about 60 ms for the `ruby -e` variant. Probably not much if seen individually, but it adds up in the grand scheme of things and further contributes to the increasing number of commands we run in our shell setup code before we start executing any of Homebrew's commands.

I didn't want to suggest premature optimization, but my gut feeling was that the impact would be noticeable, particularly because it's affecting every single Homebrew installation, even the 90%+ users that have a perfectly fine system Ruby as they are on a sufficiently modern ~~OS X~~ macOS.